### PR TITLE
coqPackages: etc

### PR DESCRIPTION
**Pull request by CohenCyril**
- put `findlib` in `nativeBuildInputs` of `mkCoqDerivation` to make sure `coq` packages find their ocaml plugin dependencies,
- use `propagatedBuildInputs` to make sure ocaml plugin dependencies are in path,
- updated `coqPackage.heq` (broken url),
- fixed use of `DESTDIR` and `COQMF_COQLIB` in mkCoqDerivation,
- adding `COQCORELIB` environement variable to put ocaml plugin files in the right place,
- make `metaFetch` available from `coqPackages`

following up https://github.com/NixOS/nixpkgs/pull/172806

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by CohenCyril** [comment]
@vbgl I reopened https://github.com/NixOS/nixpkgs/pull/161977 as such.
I do not understand the failure of why3...

**Comment by CohenCyril** [comment]
@GrahamcOfBorg build coqPackages

**Comment by CohenCyril** [comment]
@vbgl `why3` works if I force `ocamlPackages` to be `coq.ocamlPackages`...
So I believe the problem comes from incompatible versions of `findlib` and/or `num` at use between `coq` and `why3`.
I have no clue what is the best way to solve this problem in a clean way.

**Comment by CohenCyril** [comment]
Does it mean I should remove `findlib` from `propagatedNativeBuildInputs`  and `camlp5`, `num` and `zarith` from `propagatedBuildInputs` in `coq` derivation, and move it to (respectively) `nativeBuildInputs` and `buildInputs` in `mkCoqDerivation`?

**Comment by CohenCyril** [comment]
> @vbgl `why3` works if I force `ocamlPackages` to be `coq.ocamlPackages`... So I believe the problem comes from incompatible versions of `findlib` and/or `num` at use between `coq` and `why3`. I have no clue what is the best way to solve this problem in a clean way.

I'm a bit surprised, because the hook in `findlib` derivation places and expects plugins and libs to be in a directory determined by `ocaml` version :-/ so I wonder how this could happen.

**Comment by CohenCyril** [comment]
> > @vbgl `why3` works if I force `ocamlPackages` to be `coq.ocamlPackages`... So I believe the problem comes from incompatible versions of `findlib` and/or `num` at use between `coq` and `why3`. I have no clue what is the best way to solve this problem in a clean way.
> 
> I'm a bit surprised, because the hook in `findlib` derivation places and expects plugins and libs to be in a directory determined by `ocaml` version :-/ so I wonder how this could happen.

I think I understood the problem: two concurrent versions of `findlib` are in the environement.

**Comment by CohenCyril** [comment]
@GrahamcOfBorg build why3

**Comment by CohenCyril** [comment]
@GrahamcOfBorg build why3

**Comment by CohenCyril** [comment]
@Zimmi48 can I have another pair of eyes?

File: pkgs/development/ocaml-modules/elpi/default.nix [hunk 1]
```diff
@@ -1,12 +1,13 @@
-{ stdenv, lib, fetchzip, buildDunePackage, camlp5
+{ stdenv, lib, fetchzip, bash
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/elpi/default.nix:1]:
```suggestion
{ lib
```

File: pkgs/development/coq-modules/compcert/default.nix [hunk 2]
```diff
@@ -26,8 +26,9 @@ let compcert = mkCoqDerivation rec {
     "3.10".sha256 = "sha256:19rmx8r8v46101ij5myfrz60arqjy7q3ra3fb8mxqqi3c8c4l4j6";
   };
 
+  mlPlugin = true;
   nativeBuildInputs = [ makeWrapper ];
-  buildInputs = with ocamlPackages; [ ocaml findlib menhir menhirLib ] ++ [ coq coq2html ];
+  buildInputs = with coq.ocamlPackages; [ menhir menhirLib ] ++ [ coq2html ];
```
Comment (vbgl) [inline pkgs/development/coq-modules/compcert/default.nix:31]: Why these changes? CompCert is not a plugin for Coq.

Moreover, I want to compile CompCert with OCaml 5.0 even if Coq is stuck with OCaml 3.12.

Comment (CohenCyril) [reply pkgs/development/coq-modules/compcert/default.nix:31]: I think this is a side effect of my previous mistake, I will fix ASAP. Thanks for your eagle eye.

Comment (CohenCyril) [reply pkgs/development/coq-modules/compcert/default.nix]: fixed

File: pkgs/development/ocaml-modules/elpi/default.nix (hunk 1 above)
Comment (CohenCyril) [reply pkgs/development/ocaml-modules/elpi/default.nix]: fixed

**Review by vbgl — APPROVED** [review]

File: pkgs/applications/science/logic/coq/default.nix [hunk 3]
```diff
@@ -82,7 +82,8 @@ self = stdenv.mkDerivation {
 
   passthru = {
     inherit coq-version;
-    inherit ocamlPackages ocamlBuildInputs ocamlNativeBuildInputs;
+    inherit ocamlPackages ocamlNativeNuildInputs;
```
Comment (Artturin) [inline pkgs/applications/science/logic/coq/default.nix:85]: typo, fix in https://github.com/NixOS/nixpkgs/pull/207995
